### PR TITLE
improvement: Unify connecting to bazelbsp with other build servers

### PR DESCRIPTION
**Pull request by jkciesluk**

### CODE REVIEW

File: metals/src/main/scala/scala/meta/internal/metals/MetalsEnrichments.scala [hunk 1]
```diff
@@ -400,6 +401,11 @@ object MetalsEnrichments
         workspace.resolve(Directories.readonly).toNIO
       )
 
+    def isInTmpDirectory(workspace: AbsolutePath): Boolean =
```
Comment (jkciesluk) [inline metals/src/main/scala/scala/meta/internal/metals/MetalsEnrichments.scala:404]: In Bazel I was getting errors like 

```
ERROR: no such target '//:.metals/.tmp/Main-5012603866797659764.scala': target '.metals/.tmp/Main-5012603866797659764.scala' not declared in package '' defined by /.../BUILD; 
however, a source file of this name exists.  (Perhaps add 'exports_files([".metals/.tmp/Main-5012603866797659764.scala"])' to /BUILD?)
```

File: metals/src/main/scala/scala/meta/internal/metals/MetalsEnrichments.scala [hunk 2]
```diff
@@ -381,8 +381,9 @@ object MetalsEnrichments
     }
 
     def isDependencySource(workspace: AbsolutePath): Boolean =
-      (isLocalFileSystem(workspace) &&
-        isInReadonlyDirectory(workspace)) || isJarFileSystem
+      isLocalFileSystem(workspace) &&
+        (isInReadonlyDirectory(workspace) || isInTmpDirectory(workspace)) ||
```
Comment (kasiaMarek) [inline metals/src/main/scala/scala/meta/internal/metals/MetalsEnrichments.scala:385]: `.metals/.tmp/Main-...` aren't actually dependency sources, right? Maybe we could add a new method `isDependencySourceOrTmp`?

File: metals/src/main/scala/scala/meta/internal/metals/Indexer.scala [hunk 3]
```diff
@@ -121,8 +121,10 @@ final case class Indexer(
 
     bspSession() match {
       case None =>
-        scribe.warn("No build session currently active to reload.")
-        Future.successful(BuildChange.None)
+        scribe.warn(
+          "No build session currently active to reload. Attempting to reconnect."
+        )
+        reconnectToBuildServer()
```
Comment (kasiaMarek) [inline metals/src/main/scala/scala/meta/internal/metals/Indexer.scala:127]: If this now a part of a happy path, I don't think we should emit a warning here.

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala [hunk 4]
```diff
@@ -2115,19 +2095,20 @@ class MetalsLspService(
           quickConnectToBuildServer,
         )
       case None =>
-        Future.successful(BuildChange.None)
+        quickConnectToBuildServer()
```
[the remark this answers is not in the document]
Comment (jkciesluk) [reply metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala:2098]: It's a `.bloop` only situation, I'm not sure if anyone actually uses it, but I didn't want to rewrite all of our tests.

**Review by tgodzik — APPROVED** [review]

File: metals/src/main/scala/scala/meta/internal/builds/ScalaCliBuildTool.scala [hunk 5]
```diff
@@ -63,6 +63,7 @@ class ScalaCliBuildTool(
 
   override def possibleBuildServerNames = ScalaCli.names.toList
 
+  override def isAutoConnectable: Boolean = true
```
[the remark this answers is not in the document]
Comment (jkciesluk) [reply metals/src/main/scala/scala/meta/internal/builds/ScalaCliBuildTool.scala]: You're right, I've removed this method and used `isBspGenerated` instead

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala [hunk 6]
```diff
@@ -2058,7 +2051,10 @@ class MetalsLspService(
       buildTool,
       chosenBuildServer,
     )
-  } yield buildChange
+  } yield {
+    buildServerPromise.trySuccess(())
```
[the remark this answers is not in the document]
Comment (jkciesluk) [reply metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala:2041]: In `quickConnectToBuildServer` it happens even if build server is not auto connectable, so I've used similar logic here.

File: metals/src/main/scala/scala/meta/internal/metals/Indexer.scala (hunk 3 above)
Comment (jkciesluk) [reply metals/src/main/scala/scala/meta/internal/metals/Indexer.scala:127]: It's part of the happy path only on initialize. I think its best to leave the warning here, its still unhappy path is something breaks.

**Review by kasiaMarek — APPROVED** [review]
Overloading `slowConnectToBuildServer` is no longer needed to you can probably merge those methods to cleanup a little further. Otherwise, looks good, nice simplification of the logic 👍 .

**Pull request merged**